components/amara/amara.app.mjs

```diff
@@ -0,0 +1,619 @@
+    }) {
+      if (!lastResourceStr) {
+        return {
+          resources: resources.concat(nextResources),
+          resourceNotFound: true,
+        };
+      }
+
+      const lastResourceIndex =
+        resources.findIndex((resource) => JSON.stringify(resource) === lastResourceStr);
+
+      const resourceNotFound = lastResourceIndex < 0;
+
+      return {
+        resources: resources.concat(
+          resourceNotFound
+            ? nextResources
+            : nextResources.slice(0, lastResourceIndex),
+        ),
+        resourceNotFound,
+      };
+    },
+    /**
+     * paginateResources always gets the latest resources from the API.
+     * @param {Object} args - all arguments to pass to the paginateResources method
+     * @param {Function} args.resouceFn - the name of the resource function to call
+     * @param {Object} args.resourceFnArgs - the arguments object to pass to the resource function
+     * @param {number} args.offset - the offset to start at
+     * @param {number} args.limit - the number of resources to get per page
+     * @param {number} args.max - the maximum number of resources to get
+     * @param {Function} args.callback - the function to call for each resource
+     * @param {string} args.lastResourceStr - the last resource string in cache to validate against
+     *  This parameter is only passed in from sources
+     * @returns {Promise<Array>} - Promise that resolves to an array of resources,
+     *  the first element is the last resource string in cache
+     */
+    async paginateResources({
+      resourceFn,
+      resourceFnArgs,
+      offset,
+      limit = constants.DEFAULT_PAGE_LIMIT,
+      max = constants.DEFAULT_MAX_ITEMS,
+      callback,
+      lastResourceStr,
+    }) {
+      let resources = [];
+      let lastUrl;
+      let nextResponse;
+      let resourceNotFound = true;
+
+      do {
+        nextResponse = await resourceFn({
+          ...resourceFnArgs,
+          url: lastUrl,
+          params: {
+            ...resourceFnArgs.params,
+            limit,
+            offset,
+          },
+        });
+
+        if (!nextResponse) {
+          throw new Error("No response from the Amara API.");
+        }
+
+        const nextResources = nextResponse.objects;
+
+        ({
+          resourceNotFound,
+          resources,
+        } = this.processResources({
+          resources,
+          nextResources,
+          lastResourceStr,
+        }));
+
+        if (nextResponse.meta.next) {
+          lastUrl = nextResponse.meta.next;
+        }
+
+        if (callback) {
+          nextResources.forEach(callback);
+        }
+
+        // if resourceNotFound is true, then keeps looping until found
+        // if nextResponse?.meta.next is not null, then there are more resources to fetch
+        // if resources.length < max is the max number of resources to be fetched
+      } while (resourceNotFound && nextResponse?.meta.next && resources.length < max);
```

Comment:
Ok for your first paragraph, what will happen in the second run is, if there is more events created I'm actually taking into acount those since the sorting is in ascending order, so latest items will show up until `lastResourceStr` is found.

For your second paragraph, yeah I think is a good idea to add the `deploy` hook, so in that way I'm not calling all those resources again when updating the component.

For the third paragraph I'm not used to work with generators, I've always thought you can do everything with closures and functions and generators are just syntactic sugar in javascript. I'm aware there is some advantages in terms of memory and time performance in the browser context in this case. Either way is just a personal preference so I'm going to be switching to `Generators` in favor of the guidlines.

Reply:
I think using a callback function to process events, like the sources do currently, should work about the same as generators. But async generators allow us to express asynchronous code in a sequential/synchronous way that is easier to follow and understand... in the way promises or `async`/`await` improve readability and maintainability when replacing callbacks.